File: code/ii-core/src/main/java/com/owlunit/core/algo/Recommender.java
package com.owlunit.core.algo;

import com.owlunit.core.ii.Ii;
import com.owlunit.core.ii.IiDao;

import java.util.Map;

/**
 * Many Monkeys
 *
 * @author Anton Chebotaev
 */
public interface Recommender {

    /**
     * Compares two items based on a secret metric
     * Actually calls {@see compareItemsMaps(Map<InformationItem, Double> a, Map<InformationItem, Double> b)}
     *
     * @param left  item
     * @param right item
     * @return value of metric
     */
    double compareItems(Ii left, Ii right);

    /**
     * Compares two maps of components as if it would be InformationItems based on secret metric
     *
     * @param a map
     * @param b map
     * @return value of metric
     */
    double compareItemsMaps(Map<Ii, Double> a, Map<Ii, Double> b);

    /**
     * Searches similar items to provided one. Based on a comparation metric
     * Actually calls {@see getMostLike(Map<InformationItem, Double> items)}
     *
     * @param item that used as a query
     * @param dao  to perform fast loads
     * @return map os similar items with comparation results sorted by value of comparation
     */
    Map<Ii, Double> getMostLike(Ii item, IiDao dao);

    /**
     * Searches similar items to map of components as if it would be components of Ii
     * Based on a comparation metric
     *
     * @param items used as a query
     * @param dao   to perform fast loads
     * @return map os similar items with comparation results sorted by value of comparation
     */
    Map<Ii, Double> getMostLike(Map<Ii, Double> items, IiDao dao);
}

File: code/ii-core/src/main/scala/com/owlunit/core/ii/Recommender.scala
package com.owlunit.core.ii

import collection.mutable.{Set => MutableSet, Map => MutableMap}

/**
 * @author Anton Chebotaev
 *         Owls Proprietary
 */

object Recommender {

  def compareMaps(a: Map[Ii, Double], b: Map[Ii, Double]): Double = {
    val union = MutableSet[Ii]() ++ a.keys ++ b.keys

    if (union.size == 0)
      return 0.0

    val aOverall = a.values.foldLeft(0.0)(_ + _)
    val bOverall = b.values.foldLeft(0.0)(_ + _)

    var min = 0.0

    for (item <- union) {
      val aWeight = a.getOrElse(item, 0.0) / aOverall
      val bWeight = b.getOrElse(item, 0.0) / bOverall

      min += aWeight min bWeight;
    }

    min * 100
  }

}


class Recommender(dao: IiDao) {

  def compareItems(a: Ii, b: Ii): Double = Recommender.compareMaps(
    dao.getIndirectComponents(a),
    dao.getIndirectComponents(b))

  def getMostLike(a: Ii): Map[Ii,  Double] = {
    getMostLike(a.components.get) - a -- a.components.get.keys
  }

  //TODO Anton Chebotaev - IMPLEMENT
  def getMostLike(map: Map[Ii,  Double]): Map[Ii,  Double] = map

}